docs(design): add latency definition, records service and visualization (#106)
tier4/CARET_doc: docs/design/latency_definitions/node.md

# Node

Node latency is elapsed time from message reception to message transmission in a node.

$$
l_{node} = t_{transfer} - t_{receive}
$$

As transmission is replaced by publish and reception is subscription in the context of ROS 2, the equation is translated into the next one.

$$
l_{node} = t_{pub} - t_{sub}
$$

## Message context

A node receives a message, processes it, and then publishes it to a subsequent node.  
The dependency between the subscribed and published messages is used to define node latency.  
In the following sections explain the message dependency (message context), which is the concept of node latency.

For example, consider the following callback.

```c++
auto subsription_callback = [](&msg){
        msg_ = f(msg);
        pub.publish(msg_);
}
```

In this case, the received message is processed and published immediately.  
The dependency of the message at this time is described in chronological order and is expressed as follows.

![message context](../../imgs/message_context.drawio.png)

> Here, the elapsed time from subscription is defined as node latency.  

Reviewer: Sorry, but I had taken a simple mistake.
```suggestion
Here, the elapsed time from subscription to publish is defined as node latency.  
```

# Node

> Node latency is the time difference between when a callback takes ownership and when it abandons.

Reviewer: ```suggestion
Node latency is elapsed time from message subscription to message publish.
```

$$
l_{node} = t_{transfer} - t_{receive}
$$

> In particular, if transfer of ownership is subscribe and receive of ownership is publish, it is defined as follows

Reviewer: ```suggestion
As transmission is replaced by publish and reception is subscription in the context of ROS 2, the equation is translated into the next one.
```

$$
l_{node} = t_{pub} - t_{sub}
$$

> See [Communication](./communication.md) for the reason ownership is used to define latency.

Reviewer: Could you remove this sentence?

## Message context

A node receives a message, processes it, and then publishes it to a subsequent node.  
The dependency between the subscribed and published messages is used to define node latency.  
In the following sections explain the message dependency (message context), which is the concept of node latency.

For example, consider the following callback.

```c++
auto subsription_callback = [](&msg){
        msg_ = f(msg);
        pub.publish(msg_);
}
```

In this case, the received message is processed and published immediately.  
The dependency of the message at this time is described in chronological order and is expressed as follows.

![message context](../../imgs/message_context.drawio.png)

> Here, the time difference between Subscribe and Publish can be calculated as node latency.  

Reviewer: ```suggestion
Here, the elapsed time from subscription is defined as node latency.  
```

In this way, node latency can be calculated when message dependencies are defined.

> 6

Reviewer: ```suggestion

```

In the previous example, we presented a case where a node is subscribed and immediately published.

> In actuality, the Message context can be quite complex, because node processing can be defined by the developer.

Reviewer: ```suggestion
Actually, the message context might be quite complex when dependency between the incoming message and the outgoing one is complicated on multiple callbacks.
```

> In actuality, the Message context can be quite complex, because node processing can be defined by the developer.

Reviewer: ```suggestion
Actually, the message context can be quite complex because data processing in a node is freely defined by an application developer.
```

![complex message context](../../imgs/message_context_complex.drawio.png)

> Each case is described below.

Reviewer: ```suggestion
Examples of complex message context are described below.
```

- Buffering, for example, is a buffer delay process.
- Multiple use is for moving average processing, for example. There are multiple candidates for node latency.
- Unused is a message that has not been published and has not been used. This is evaluated as a kind of message drop.

> - Cross can occur in cases where messages are retrieved based on the timestamp of the message rather than the system time.

Reviewer: How about renaming as blow to align participle of words;

- Buffering -> Buffered
- Multiple use -> Multiple inputted
- Unused -> Unused (not changed)
- cross -> Crossed

Reviewer: Could you change the corresponding figure?

Author: addressed  [58b88f0](https://github.com/tier4/CARET_doc/pull/106/commits/58b88f078e12f475c24b0efda0feb1b235e99772)

> In either case, it is difficult to automatically determine the message context.

Reviewer: ```suggestion
In any of those cases, it is difficult to automatically determine the message context.
```

> In some cases this can be covered by Configuration.

Reviewer: ```suggestion
the message context are given for some cases of them as shown in configuration.
```

For more information, see [Configuration](. /... /configuration/index.md) for details.

> If Configuration does not cover the situation, we are considering a mechanism to allow users to describe message dependencies as a future plan.

Reviewer: ```suggestion
To enlarge coverage the others of complex cases, we are considering a mechanism to allow users to describe message dependencies in their user codes.
```

> For details, please refer to [TILDE](. /software_architecture/tilde.md) for details.

Reviewer: ```suggestion
[TILDE](. /software_architecture/tilde.md) is one of good candidates to achieve it.
```
